### ISSUE TITLE
.NET 8 developers can verify more APIs for correct usage to speed up their development

### ISSUE THREAD

**Issue by buyaa-n** [dotnet/runtime#78442]
With the Roslyn Analyzers we built in .NET 5, 6 and 7, we've found that developers benefit from analyzers that verify correct usage of .NET Libraries APIs. We plan to put more effort on adding valuable analyzers in .NET 8, we will continue working with the community to foster contributions into our collection of analyzers and code fixers. 

Community contributors: If you'd like to work on one of the issues marked as "up for grabs", please add a comment directly on that issue asking to get it assigned to you. Please check the [getting started guide](https://github.com/dotnet/roslyn-analyzers/blob/master/docs/NetCore_GettingStarted.md) for reference about how to start and what's need to be done.

### Approved analyzers 
| No | Status | Issue | Estimate |
|-|--------|-------| -------|
| 1 | :heavy_check_mark: by @CollinAlpert | [Prefer Dictionary<K, V>.TryAddValue(key) over guarded Add(key)](https://github.com/dotnet/runtime/issues/33799) | Medium |
| 2 | :heavy_check_mark: by @CollinAlpert | [Prefer .Length/Count/IsEmpty over Any()](https://github.com/dotnet/runtime/issues/75933) | Medium |
| 3 | :heavy_check_mark: by @fowl2 | [Do not use OfType<T>() with impossible types](https://github.com/dotnet/runtime/issues/33770) | Small |
| 4 | :heavy_check_mark: by @steveberdy | [Extract array of const to static readonly field](https://github.com/dotnet/runtime/issues/33794) | Medium |
| 5 | [in PR](https://github.com/dotnet/roslyn-analyzers/pull/4877) by @psxvoid | [Constructor parameters should match property names](https://github.com/dotnet/runtime/issues/33796) | Medium |
| 6 | [in PR](https://github.com/dotnet/roslyn-analyzers/pull/5506) by @NewellClark | [Prevent wrong usage of string Split](https://github.com/dotnet/runtime/issues/47927) | Medium |
| 7 | [in PR](https://github.com/dotnet/roslyn-analyzers/pull/5548) by @NewellClark | [Prefer static ReadOnlySpan<byte> properties over static readonly byte[] fields](https://github.com/dotnet/runtime/issues/33780) | Medium |
| 8 | [in PR](https://github.com/dotnet/roslyn-analyzers/pull/5454) by @NewellClark | [Do not pass Utf8JsonReader by value](https://github.com/dotnet/runtime/issues/33772) | Medium |
| 9 | :heavy_check_mark: @geeknoid | [Recommend use of concrete types to maximize devirtualization potential](https://github.com/dotnet/runtime/issues/51193) | Medium |
| 10 | :heavy_check_mark: @stephentoub | [Convert argument null checks to ArgumentNullException.ThrowIfNull](https://github.com/dotnet/runtime/issues/68326) | Medium |
| 11 | :heavy_check_mark: @Youssef1313 | [Using StartsWith instead of IndexOf == 0](https://github.com/dotnet/runtime/issues/78608) | Medium |
| 12 | @NewellClark | [Avoid implicitly allocating params arrays in loops](https://github.com/dotnet/runtime/issues/33793) | Medium |
| 13 | :heavy_check_mark: @carlossanlop | [Recommend string.ToLower{Invariant}().Contains/IndexOf/StartsWith("lowercasestring") be replaced by case-insensitive comparison](https://github.com/dotnet/runtime/issues/78606) | Medium |
| 14 | up for grabs | [Usage of ToString in logging source generator method call sites](https://github.com/dotnet/runtime/issues/78402) | Medium |
| 15 | :heavy_check_mark: @carlossanlop | [string.ToLower() == otherString.ToLower()](https://github.com/dotnet/runtime/issues/78607) | Small |
| 16 | :heavy_check_mark: @mrahhal | [Recommend string.StartsWith('c') instead of string.StartsWith("c")](https://github.com/dotnet/runtime/issues/78392) | Medium |
| 17 | up for grabs | [Replacing direct LoggerExtensions.Log* usage with source-generated logging methods](https://github.com/dotnet/runtime/issues/78406) | Medium |
| 18 | up for grabs | [Prefer ValueTuple over Tuple](https://github.com/dotnet/runtime/issues/33791) | Medium |
| 19 | :heavy_check_mark: @mpidash | [Report inefficient use of sets](https://github.com/dotnet/runtime/issues/85490) | Small |
| 20 | up for grabs | [Introduce an analyzer to recommend the more efficient string splitting features in .NET 8](https://github.com/dotnet/runtime/issues/85487) | Medium |
| 21 | partially implemented @stephentoub | [Recommend upgrading to CompositeFormat for enhanced performance](https://github.com/dotnet/runtime/issues/85525) | Large |
| 22 | up for grabs | [Prefer Dictionary.Remove(key, out value) over Dictionary.this[key], followed by Dictionary.Remove(key)](https://github.com/dotnet/runtime/issues/33800) | Medium |
| 23 | up for grabs | [Hide Thread.VolatileRead and Thread.VolatileWrite](https://github.com/dotnet/runtime/issues/27997) | Medium |
| 24 | in PR by @mpidash | [Simplify calls to string.Substring and Memory/Span.Slice](https://github.com/dotnet/runtime/issues/68946) | Small |
| 25 | @mpidash | [Detect the problem of using a System.Type overload instead of the generic overload](https://github.com/dotnet/runtime/issues/68243) | Medium |
| 26 | up for grabs | [Treat calls to ROS<char> op_Equality as warning](https://github.com/dotnet/runtime/issues/54794) | Small |
| 27 | up for grabs | [Detect non-cancelable Task.Delay passed to Task.WhenAny](https://github.com/dotnet/runtime/issues/33805) | Medium |
| 28 | @tannergooding | [Upgrade from platform specific to cross platform intrinsics](https://github.com/dotnet/runtime/issues/82486) | Small |
| 29 | @tannergooding | [Suggest more optimal SIMD patterns where applicable](https://github.com/dotnet/runtime/issues/82488) | Medium |
| 30 | up for grabs | [Adjust/simplify code for numeric IntPtr](https://github.com/dotnet/runtime/issues/74518) | Medium |
| 31 | up for grabs | [Analyzer/fixer for converting Stream.Read calls to ReadAtLeast and ReadExactly](https://github.com/dotnet/runtime/issues/69159) | Medium |
| 32 | up for grabs | [Exceptions thrown inside async methods should be wrapped by Task.FromException](https://github.com/dotnet/runtime/issues/70905) | Medium |
| 33 | @CollinAlpert  | [Make types declared in an executable internal](https://github.com/dotnet/runtime/issues/78388) | Small |
| 34 | up for grabs | [Unawaited / returned tasks created with a disposable in using block for that disposable](https://github.com/dotnet/runtime/issues/78394) | Large |
| 35 | @pedrodeandrade | [Async methods that don't take a CancellationToken](https://github.com/dotnet/runtime/issues/78397) | Small |
| 36 | up for grabs | [Use IndexOfAnyValues instead of inlined or cached array](https://github.com/dotnet/runtime/issues/78587) | Small |
| 37 | :heavy_check_mark: @Jozkee | [Add an analyzer that warns on single-use JsonSerializerOptions instances](https://github.com/dotnet/runtime/issues/65396) | |

### Other analyzer proposals that need to be prepared for API review
 1. https://github.com/dotnet/runtime/issues/31354
 2.  https://github.com/dotnet/runtime/issues/71698
 3. https://github.com/dotnet/runtime/issues/35981
 4. https://github.com/dotnet/runtime/issues/45215
 5. https://github.com/dotnet/runtime/issues/50389
 6. https://github.com/dotnet/runtime/issues/64009
 7. https://github.com/dotnet/runtime/issues/64545
 8. https://github.com/dotnet/runtime/issues/68962
 9. https://github.com/dotnet/runtime/issues/77625
 10. https://github.com/dotnet/runtime/issues/78398

CC @stephentoub @jeffhandley @ericstj @carlossanlop

**Comment by aradalvand** [comment]
The `.Any()` => `.Length != 0` warning makes no sense whatsoever.
`.Any()` should check under the hood whether the incoming `IEnumerable` is in fact an array, in which case it should do `.Length != 0` automatically, LINQ method already do this sort of thing extensively, as far as I know.

We shouldn't be telling users to change their code in these cases, `.Any()` is undeniably clearer and more readable than `.Length != 0`.

**Comment by stephentoub** [comment]
> .Any() should check under the hood whether the incoming IEnumerable is in fact an array, in which case it should do .Length != 0 automatically. LINQ methods already do this sort of optimization-based-on-the-type extensively, as far as I know.

It does, and it's still more costly than the caller just doing it itself.

**Comment by aradalvand** [comment]
> It does, and it's still more costly than the caller just doing it itself.

Good to know, but then again if `Any()` merits an analyzer/warning, so does every other LINQ method that does type-checking under the hood — in cases where the caller knows the type of the collection, no? Why make an exception for `.Any()`?

**Comment by mrahhal** [comment]
> I don't understand why .Any() is being singled out here.

Likely because it's the most common usage in linq.

But I agree that, at least, this probably should have a lower diagnostic level than Suggestion by default. Most code won't benefit from this particular optimization, any code that is sensitive enough to benefit would probably be already aware of this. I also feel that `Any()` has become more readable overtime than `Length/Count == 0`. Maybe because it's a common interface that doesn't need to change when you're switching impls from arrays to collections and vice versa as you're developing. Having largely irrelevant squiggles randomly popping up isn't pretty, I always end up silencing this particular family of diags.

**Comment by Bellarmine-Head** [comment]
Just found this (via Nick Chapsas).

Re. https://github.com/dotnet/runtime/issues/78606 - it's a pity that the word "ordinal" does not appear even once... especially wrt ignore-case.  I think it should have featured in the discussion.

[This table](https://github.com/dotnet/runtime/issues/78606#issuecomment-1564938950) looks wrong to me.  In every case, the assumption is that the string is linguistic, or is "linguistically relevant".

I refer you to:- https://learn.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings

Specifially I refer you to: https://learn.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings#choosing-a-stringcomparison-member-for-your-method-call

For over 20 years, my case-insensitve string comparisons have been wrt

* Case-insensitive internal identifiers.
* Case-insensitive identifiers in standards such as XML and HTTP.
* File paths.
* Registry keys and values.
* Environment variables.
* Resource identifiers (for example, handle names).
* Case-insensitive security-related settings.

ergo... `StringComparison.OrdinalIgnoreCase`.

The number of times I have had to worry about case-insensitivity in linguistic data that's input by the user or displayed to the user: **zero.**

The more common case (by far) for case-insensitive string comparisons is for non-linguistic strings.  And getting it right for non-linguistic strings is important for security.

In every single case [in the table](https://github.com/dotnet/runtime/issues/78606#issuecomment-1564938950), I'd go for `OrdinalIgnoreCase` as the default analyzer recommendation, with the caveat that "if the string is linguistic, go for culture-ignore-case".

**Comment by mirmostafa** [comment]
Instead of sending warning when using `.Any()` in arrays and list, why not to add the following bunch of code to the .Net sources:
```CSharp
    public static bool Any<T>(this IList<T> source) =>
        source.Count != 0;
    public static bool Any<T>(this T[] source) =>
        source.Length != 0;
    public static bool Any<T>(this ICollection source) =>
        source.Count != 0;
```

**Comment by buyaa-n** [comment]
As .NET 8 development is wrapping up closing this issue. Thank you, all the contributors, for the hard work and hope you will continue contributing into new analyzers in .NET 9 as we will continue the effort in .NET 9, analyzers that haven't completed with this issue will be moved to a new issue for .NET 9.

**Comment by Bellarmine-Head** [comment]
@buyaa-n 

I am concerned that analyzers https://github.com/dotnet/runtime/issues/78606 and https://github.com/dotnet/roslyn-analyzers/pull/6720 and possibly others that are related have been implemented incorrectly, and I left multiple comments 3 weeks ago explaining why - all of which have been ignored.

I would like to see some comment on these issues.  We don't want to wrap things up in an incorrect state.

My concerns are expressed here:-

* https://github.com/dotnet/runtime/issues/78442#issuecomment-1646526967
* https://github.com/dotnet/roslyn-analyzers/pull/6720#issuecomment-1646528348
* https://github.com/dotnet/runtime/issues/89740#issuecomment-1659726231

**Comment by buyaa-n** [comment]
@Bellarmine-Head I assume your comments related to [this issue](https://github.com/dotnet/runtime/issues/89740) which is not closed and still have 8.0 milestone. I expect @carlossanlop would take care of it within 8.0 timeframe.

**Comment by Bellarmine-Head** [comment]
@buyaa-n I added my comments to https://github.com/dotnet/runtime/issues/89740 because you asked me to.  I'm not sure what the issue is all about tbh (!) nor how it relates to my main concerns over issues:-

* https://github.com/dotnet/runtime/issues/78606
* https://github.com/dotnet/roslyn-analyzers/pull/6720

... which are closed/merged, but I believe have been implemented incorrectly for the reasons that I have stated multiple times.

It would be great if @carlossanlop and other experts could comment on these concerns.

**Comment by Bellarmine-Head** [comment]
@buyaa-n 

To expand on my concerns...

Let us take the example of https://github.com/dotnet/runtime/issues/78606 which was opened by @stephentoub 

It's fixed by https://github.com/dotnet/roslyn-analyzers/pull/6662

Nowhere in the discussion in those two pages is the word "ordinal".  This was the first red flag for me.

Now, to take a small example, it seems to me that the analyzer will recommend that where you previously did:-

```
str.ToLowerInvariant().IndexOf(substr)
```

you should now do:-

```
str.IndexOf(substr, StringComparison.InvariantCultureIgnoreCase)
```

This seems reasonable.  What you had before, you have afterwards, except arguably in a better form.

But you have overlooked something.  [This section](https://learn.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings#choosing-a-stringcomparison-member-for-your-method-call) in this [long-standing (and correct) article](https://learn.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings), which has been around in one form or another since 2005, is very clear on the matter.  For comparing:-

* Case-insensitive internal identifiers.
* Case-insensitive identifiers in standards such as XML and HTTP.
* File paths.
* Registry keys and values.
* Environment variables.
* Resource identifiers (for example, handle names).
* Case-insensitive security-related settings.

you should be using `StringComparison.OrdinalIgnoreCase`.

For me this boils down to:-

"Case-insensitive string comparisons should always use `StringComparison.OrdinalIgnoreCase` except in the *very rare* cases where the strings you are comparing are linguistic in some shape or form."

Examples of linguistic strings: a blog piece the user has typed in; a date-time formatted in the user's culture for end-user display.

Why you would ever want to compare linguistic strings case-insensitively, I don't know.  In 20 years, I've never done it.  Comparing (non-linguistic) identifiers, on the other hand, is something you do all day, every day.

If our analyzer (and related ones; see links in my post above) _doesn't even consider_ or mention or recommend the `StringComparison.OrdinalIgnoreCase` case, then in my view it has to be deemed as grossly deficient and represents three problems:-

* it could cause Microsoft's own programmers to do the wrong thing when comparing identifiers
* it can be argued that it's recommending non-Microsoft programmers to do the wrong thing
* it's going against https://learn.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings#choosing-a-stringcomparison-member-for-your-method-call

Now, all of this is *my opinion*.  The lack of response to these concerns, which I have written about in sundry places (see links above) concerns me.

Talk of wrapping things up for .NET 8 without addressing this issue concerns me.

I'm putting in a lot of effort here in an attempt to help... and I'm not getting much/any response from it.

**Comment by MichalPetryka** [comment]
> @buyaa-n
> 
> To expand on my concerns...
> 
> Let us take the example of #78606 which was opened by @stephentoub
> 
> It's fixed by [dotnet/roslyn-analyzers#6662](https://github.com/dotnet/roslyn-analyzers/pull/6662)
> 
> Nowhere in the discussion in those two pages is the word "ordinal". This was the first red flag for me.
> 
> Now, to take a small example, it seems to me that the analyzer will recommend that where you previously did:-
> 
> ```
> str.ToLowerInvariant().IndexOf(substr)
> ```
> 
> you should now do:-
> 
> ```
> str.IndexOf(substr, StringComparison.InvariantCultureIgnoreCase)
> ```
> 
> This seems reasonable. What you had before, you have afterwards, except arguably in a better form.
> 
> But you have overlooked something. [This section](https://learn.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings#choosing-a-stringcomparison-member-for-your-method-call) in this [long-standing (and correct) article](https://learn.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings), which has been around in one form or another since 2005, is very clear on the matter. For comparing:-
> 
> * Case-insensitive internal identifiers.
> * Case-insensitive identifiers in standards such as XML and HTTP.
> * File paths.
> * Registry keys and values.
> * Environment variables.
> * Resource identifiers (for example, handle names).
> * Case-insensitive security-related settings.
> 
> you should be using `StringComparison.OrdinalIgnoreCase`.
> 
> For me this boils down to:-
> 
> "Case-insensitive string comparisons should always use `StringComparison.OrdinalIgnoreCase` except in the _very rare_ cases where the strings you are comparing are linguistic in some shape or form."
> 
> Examples of linguistic strings: a blog piece the user has typed in; a date-time formatted in the user's culture for end-user display.
> 
> Why you would ever want to compare linguistic strings case-insensitively, I don't know. In 20 years, I've never done it. Comparing (non-linguistic) identifiers, on the other hand, is something you do all day, every day.
> 
> If our analyzer (and related ones; see links in my post above) _doesn't even consider_ or mention or recommend the `StringComparison.OrdinalIgnoreCase` option, then in my view it has to be deemed as grossly deficient and represents three problems:-
> 
> * it could cause Microsoft's own programmers to do the wrong thing when comparing identifiers
> * it can be argued that it's recommending non-Microsoft programmers to do the wrong thing
> * it's going against https://learn.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings#choosing-a-stringcomparison-member-for-your-method-call
> 
> Now, all of this is _my opinion_. The lack of response to these concerns, which I have written about in sundry places (see links above) concerns me.
> 
> Talk of wrapping things up for .NET 8 without addressing this issue concerns me.
> 
> I'm putting in a lot of effort here in an attempt to help... and I'm not getting much/any response from it.

The issue would be that changing stuff to Ordinal in such case would be a visible behavioural change, which is something that analyzers shouldn't always promote.

**Comment by Bellarmine-Head** [comment]
For case-sensitive string comparisons, we do

```
if (str1 == str2)
```

and this gives an ordinal comparison (i.e. `StringComparison.Ordinal`).

So these analyzers are concerned with case-insensitive comparisons for the most part.

Recommending to use `String.Equals` instead of `if (stringOne.ToUpperInvariant() == stringTwo.ToUpperInvariant())` for performance reasons, while retaining maximum backwards compatibility (no change in behaviour) via use of `StringComparison.InvariantCultureIgnoreCase` is a worthy thing.

But a higher, better, nobler and more worthy thing is to recommend the use of `StringComparison.OrdinalIgnoreCase` for correct operation and good security - in a word: "safety".

Safety of implementation beats out perf any day of the week.

The downside is that ppl have to check that there is no change in behaviour, I get that.
